### DISCUSSION TITLE
For loop syntax

### DISCUSSION THREAD

**Discussion by jlp765** [felix-lang/felix#182] [Ideas]
For newbies, there is an unnecessary cognitive hurdle required to get for loops correct.

When do you use a var and when not to?  
The question is: **Could this be eliminated, to make it easier to use Felix?**

**Example**:
(The 3rd and 5th loops in the following code)  
When do you need to add the `var` to the for loop?  This is unnecessary thinking on the part of the programmer, and results in extra bug fixing when coding with the language.
If Felix could determine that the variable doesn't exist and make it a local var to the `for` loop, it would make things simpler and easier (no var required in the syntax.  Define it outside the loop, or let Felix define it for you for looping purposes).
```
var lst = list(0,1,2,3,4,5,6,7,8,9,10);

var i = 0;
var llen = lst.len.int - 1;

print$ "Pre-defined var - upto: ";
for i in 0 upto llen perform print$ (lst.i.str + " ");
println$ "";

print$ "Pre-defined var - in: ";
for i in lst perform print$ (lst.i.str + " ");
println$ "";

print$ "Loop var - upto: ";
for var j in 0 upto llen perform print$ (lst.j.str + " ");
println$ "";

print$ "Loop var - in: ";
for k in lst perform print$ (lst.k.str + " ");
println$ "";

print$ "Loop var - in: ";
for var m in lst perform print$ (lst.m.str + " ");   //  OOPS - no var here, but an easy mistake
println$ "";
```

**Comment by skaller** [comment]
Yes agreed. For low level loop constructions, you have to declare the variable outside the loop. For some constructions you can *also* declare it the loop BUT it is NOT local to the loop, it persists so the same loop again will cause a duplicate symbol error. And for some loops, the variable is auto declared and local to the loop. It's a mess.

Localising the variable is done (see the grammar) by nesting it in a procedure then calling the procedure so the variable becomes local to the procedure. The procedure call is then inlined. The nasty is that a goto inside the procedure to outside the procedure is a non-local goto until after inlining, so the inlining is mandatory.

The problem with local loop variables is that they don't survive the loop and you often want them to. On the other hand, if you export the control variable outside the loop, you can only declare the variable in that loop and it's seen everywhere.

https://github.com/felix-lang/felix/blob/master/src/packages/grammar.fdoc

around line 3600 stuff on loops. As you can see, most of the loops are implemented in the grammar using gotos.

**Comment by skaller** [comment]
# RAW LOOPS
All the raw loops are flat, meaning, you can jump into the middle of one, the body does *not* define a scope. This means the forms with `var` declare a new variable which is visible in the current scope, and which will conflict with any other variable of the same name already or subsequently declared.

### C like loops
The initial statement may assign or define the control variable. The control variable *overshoots*.
```
loop_stmt := optlabel "for" "(" stmt sexpr ";" stmt ")" stmt =>#  
loop_stmt := optlabel "for" stmt "while" sexpr ";" "next" stmt block =>#
loop_stmt := optlabel "for" stmt "until" sexpr ";" "next" stmt block =>#
```

### count ups
The control variable does *not* overshoot. 
One form to assign the control variable and two to declare it (one with and one without a specified type)
```
loop_stmt := optlabel "for" "var" sname ":" sexpr "in" sexpr "upto" sexpr block =>#  
loop_stmt := optlabel "for" "var" sname "in" sexpr "upto" sexpr block =>#
loop_stmt := optlabel "for" sname "in" sexpr "upto" sexpr block =># "(assign_incluploop)";
```

### count downs
The control variable does *not* overshoot. 
One form to assign the control variable and two to declare it (one with and one without a specified type)
```
loop_stmt := optlabel "for" sname "in" sexpr "downto" sexpr block =>#
loop_stmt := optlabel "for" "var" sname ":" sexpr "in" sexpr "downto" sexpr block =>#
loop_stmt := optlabel "for" "var" sname "in" sexpr "downto" sexpr block =>#
```
## slice ups
```
loop_stmt := optlabel "for" sname "in" sexpr ".." sexpr block =># "(define_incluploop)";
loop_stmt := optlabel "for" sname "in" sexpr "..<" sexpr block =># "(define_excluploop)";
```

# iterator
These are high level loops with a body forming a separate scope.
The control variable is not visible outside the loop.
You cannot jump into the body.
`for` is tail recursive.
`rfor` is not.
The body is embedded into a procedure which is then called, creating a local scope.
The procedure is marked `force inline` to eliminate the overhead of a call.

```
loop_stmt := optlabel "for" sname "in" sexpr block =>#
loop_stmt := optlabel "rfor" sname "in" sexpr block =># '(iterator_recursive_loop _1 _3 _5 _6)';
```
